bpf/lib/nodeport.h

```diff
@@ -1640,14 +1644,22 @@ int tail_nodeport_nat_ingress_ipv4(struct __ctx_buff *ctx)
 
 	/* At this point we know that a reverse SNAT mapping exists.
 	 * Otherwise, we would have tail-called back to
-	 * CALL_IPV4_FROM_NETDEV in the code above. The existence of the
-	 * mapping is an indicator that the packet might be a reply from
-	 * a remote backend. So handle the service reverse DNAT (if
-	 * needed)
+	 * CALL_IPV4_FROM_NETDEV in the code above.
+	 */
+#if !defined(ENABLE_DSR) || (defined(ENABLE_DSR) && defined(ENABLE_DSR_HYBRID)) ||	\
+    (defined(ENABLE_EGRESS_GATEWAY) && !defined(TUNNEL_MODE))
```

Comment:
Could you expand on this case in a code comment or in the commit? It seems like the trickiest case, yet it's relegated to a footnote :sweat_smile:  
In particular, why does `TUNNEL_MODE` make a difference here? Doesn't egress gateway traffic go through the tunnel regardless of the routing mode?

Reply:
Just to be clear - you saw the comments right below this line, but they're not sufficient?

Reply:
Yes I did. The only comment below on the egress gateway case describes what is being done, not why:

> Also redirect EgressGW reply traffic into tunnel.

Reply:
Ah - yes, we're piggy-backing on the RevDNAT code for EgressGW replies. The details are in https://github.com/cilium/cilium/blob/1b4f2a334a8f3826c925a5bf49b7fc1165f292f4/bpf/lib/nodeport.h#L1976-L1981

Yes, that's not pretty - and we should probably do a s/CILIUM_CALL_IPV4_NODEPORT_REVNAT/CILIUM_CALL_IPV4_NODEPORT_NAT_INGRESS_CONTINUE to clarify that it's also handling some non-RevDNAT cases.

Reply:
Ouch! I see. We probably need a code comment here as well; it may be enough to point to "the egress gateway special case in rev_nodeport_lb{4,6}".

Reply:
Added a comment:

> Also let rev_nodeport_lb4() redirect EgressGW reply traffic into tunnel (see there for details).